### ISSUE TITLE
Non english keyboard layout

### ISSUE THREAD

**Issue by elig0n** [Genymobile/scrcpy#37]
I have hebrew keyboard layout on both my computer and my android keyboard that is in use but it wouldn't let me pass the letters.  Any chance to fix this?

**Comment by rom1v** [comment]
> Any change to fix this?

[Probably not](https://blog.rom1v.com/2018/03/introducing-scrcpy/#handling-accented-characters).

Except if someone find a technical solution ;)

**Comment by MiracleXYZ** [comment]
I used to have the same issue but it disappeared through some "practice":
1. Turn to AOSP and type some numbers like `1234` (that works fine).
2. Type some letters. If it doesn't work, press `shift` key several times. (You may say a bunch of letters "crowding in".
3. Turn to GBoard or other keyboards. It will work fine if you are lucky :)

**Comment by wuairc** [comment]
One possible technical solution is to write a customized input method manager. [Vysor](https://chrome.google.com/webstore/detail/vysor/gidgenkbbabolejbgbpnhbimgjbffefm)(an scrcpy alternative) does this.
The input method manager can be installed and enabled by adb, without user interaction.

**Comment by AndroidDeveloperLB** [comment]
+1 on this. Wanted to write in Hebrew too, but failed.
Are you sure it's impossible to use adb to pass unicode keys ?

**Comment by vshymanskyy** [comment]
One of the possible options could be to create a custom keyboard, like `NULL keyboard` that I described here: https://github.com/Genymobile/scrcpy/issues/181, but with an ability to Input and Copy/Paste unicode. Probably not the easiest option, but at least something if there's not system API for such things.

**Comment by AndroidDeveloperLB** [comment]
I would prefer without this workaround...

**Comment by giorgiobeggiora** [comment]
+1

**Comment by giorgiobeggiora** [comment]
I strongly hope this will be fixed soon

**Comment by livin2** [comment]
https://remoboard.app/
example of this app may be help. 
pass unicode via an wireless server independent from adb server.

**Comment by AndroidDeveloperLB** [comment]
@livin2 This means I have to change my default keyboard, which means I might have less features than what I want to have.
If an app is required, I would even install one that requires root, but having less features and having to switch keyboards as an alternative... This is just a hassle...

Is there maybe a way to do it using "monkeyrunner "? Someone wrote about this here:
https://stackoverflow.com/a/20615725/878126

**Comment by livin2** [comment]
> If an app is required, I would even install one that requires root, but having less features and having to switch keyboards as an alternative... This is just a hassle...

feature like emoji mostly depend on the ime of computer.  
and you can switch keyboards with adb shell in the script where you start scrcpy:
```
adb shell ime list -s 
```
```
adb shell ime set 【IME NAME】
adb connect 【xxx】
scrcpy -s 【xxx】
```

**Comment by AndroidDeveloperLB** [comment]
Yes, still a hassle. Once I disconnect from USB, I have to switch back to the keyboard I use.
Besides, all this time that I use Scrcpy I won't have any of the features of the keyboard that I use.

**Comment by pangliang** [comment]
Refer to the Vysor method, add a custom ime to enter unicode characters
#903

**Comment by NewUserHa** [comment]
I guess it may not be that easy and other apps may implement it successfully.

hope can input and copy/paste unicode characters via scrcpy window directly.

**Comment by rom1v** [comment]
> hope can input and copy/paste unicode characters via scrcpy window directly.

Yes, this works. https://github.com/Genymobile/scrcpy#copy-paste

**Comment by NewUserHa** [comment]
I meant unicode directly.

what about putting the key event to the IME if don't implement something new?

**Comment by NewUserHa** [comment]
I think if users are playing mobile game using scrcpy, they must need type texts sometimes.
if the users are trying to type/paste chinese or japanese then scrcpy is unable to cover that needs. and I don't think scrcpy need other big features except this one.

**Comment by rom1v** [comment]
> I don't think scrcpy need other big features except this one.

and #14 :wink:

**Comment by NewUserHa** [comment]
oh, I'm just reading it while you reply.
the audio seems a bigger feature and some hard and a little important than current one. but when waking up the phone(cpu855) using scrcpy, I feel it recover from lag in the first seconds(on windows). I guess there could be problems if there's audio playing in the background.

**Comment by AndroidDeveloperLB** [comment]
Could copy-paste work method work for non-English?
I tried now using it, but it didn't paste.
It did succeed to copy it to the clipboard of the device though.
Is there maybe an Android-supported key for pasting ? If so, maybe this could be a workaround for typing...

**Comment by rom1v** [comment]
When you Ctrl+Shift+v in the scrcpy window, the computer clipboard is copied to the device clipboard (exactly, without problems with special characters).

You can then long-press in any text area on your fevice and click in Copy.

**Comment by AndroidDeveloperLB** [comment]
@rom1v I know. That's why I wrote "It did succeed to copy it to the clipboard of the device though.".
My question was if it's possible to do a real pasting, perhaps using some workaround of paste via a key.

**Comment by rom1v** [comment]
Please test #1426 (works on Android >= 7).

**Comment by JulyMorning** [comment]
any workarounds found?

**Comment by AndroidDeveloperLB** [comment]
Found another mirroring app that does it, and as was written here, it uses a keyboard app on the device:
https://en.mirroid.io/

Sadly it's only for Windows OS, and it has some annoyances. But it works for this case.

**Comment by npes87184** [comment]
POC that I implemented before.

https://github.com/npes87184/SocketIME

And its [demo](https://www.youtube.com/watch?v=s9LZpAnIf4A&feature=youtu.be).

**Comment by AndroidDeveloperLB** [comment]
@npes87184  So this is by using a new keyboard app on the device, that has to be set as default, right?

**Comment by rom1v** [comment]
@npes87184 To keep things simple on the client side, I think we could do the following:
 - put the IME service class in scrcpy-server
 - if the user pass `--use-ime` (for example), then the scrcpy-server installs itself (programatically, it has shell permissions) asynchronously (if the same version is not already installed), without delaying video stream and control (except text events, which could be queued or ignored)
 - it then enables the IME and connects to it
 - once connected, all text events received by scrcpy-server are now forwarded to the IME
 - on cleanup, scrcpy-server restore the previous IME

Tht way, the client always communicate to scrcpy-server only, which manages everything. And this doesn't delay scrcpy starting (except for text).

What do you think?

**Comment by npes87184** [comment]
> @npes87184 So this is by using a new keyboard app on the device, that has to be set as default, right?

In this POC, yes.

> @npes87184 To keep things simple on the client side, I think we could do the following:
> 
> * put the IME service class in scrcpy-server
> * if the user pass `--use-ime` (for example), then the scrcpy-server installs itself (programatically, it has shell permissions) asynchronously (if the same version is not already installed), without delaying video stream and control (except text events, which could be queued or ignored)
> * it then enables the IME and connects to it
> * once connected, all text events received by scrcpy-server are now forwarded to the IME
> * on cleanup, scrcpy-server restore the previous IME
> 
> Tht way, the client always communicate to scrcpy-server only, which manages everything. And this doesn't delay scrcpy starting (except for text).
> 
> What do you think?

I would say that this method is pretty well.
However, there is a problem about UI. How do we "make" an area to let user enter words.

**Comment by rom1v** [comment]
> However, there is a problem about UI. How do we "make" an area to let user enter words.

Oh, I've never used this kind of things. I don't really know how they work.

Isn't this area provided by the system, and give the resulting character as a "text event"?

Once we have the text event with UTF-8 content (typically for one "character", like for letters), we can inject it to the system using the IME without being limited to ASCII.

**Comment by NewUserHa** [comment]
and what about the `physical keyboard`? making user type on the device directly.
and what about adding a text control let user input texts and copy to the device by scrcpy in bachground when user type `enter`.
both ways will be better than panful installing anything.

**Comment by AndroidDeveloperLB** [comment]
Is there perhaps a way to achieve unicode key support by using accessibilityService? 
Maybe using dispatchKey() or something?
Seems to me like a better thing, instead of having to use new keyboard and reset it...

**Comment by npes87184** [comment]
> > However, there is a problem about UI. How do we "make" an area to let user enter words.
> 
> Oh, I've never used this kind of things. I don't really know how they work.
> 
> Isn't this area provided by the system, and give the resulting character as a "text event"?
> 
> Once we have the text event with UTF-8 content (typically for one "character", like for letters), we can inject it to the system using the IME without being limited to ASCII.

Oh, you misunderstand my question. My question is how do we design an area in SDL so that user can type in that area and send it to IME.

**Comment by rom1v** [comment]
https://wiki.libsdl.org/SDL_StartTextInput
https://wiki.libsdl.org/Tutorials/TextInput
(not tested)

**Comment by AndroidDeveloperLB** [comment]
@rom1v What's that?

**Comment by vip-liutao** [comment]
当手机打开投屏时，系统会自动添加wfd-uhid-device的物理键盘
这让scrcpy的input keyevent可以被输入法监听到
这很神奇，但是我不知道如何在不开启投屏的时候启用这个wfd-uhid-device物理键盘
视频测试经过地址：
https://raw.githubusercontent.com/vip-liutao/scrcpy/main/keyboard.mkv

**Comment by vip-liutao** [comment]
> 当手机打开投屏时，系统会自动添加wfd-uhid-device的物理键盘
> 这让scrcpy的输入键事件可以被输入法监听到
> 这很神奇，但是我不知道如何在不开启投屏的时候启用这个wfd -uhid-device物理键盘
> 视频测试通过地址：[https](https://raw.githubusercontent.com/vip-liutao/scrcpy/main/keyboard.mkv) : //raw.githubusercontent.com/vip-liutao/scrcpy/main/keyboard.mkv
> [](https://raw.githubusercontent.com/vip-liutao/scrcpy/main/keyboard.mkv)

@rom1v

**Comment by rom1v** [comment]
How do you enable "wfd-uhid-device" in practice?

Refs:
 - https://github.com/Genymobile/scrcpy/issues/87#issuecomment-378775565
 - https://github.com/Genymobile/scrcpy/issues/279#issuecomment-453819354

**Comment by omareg94** [comment]
When I activate raw key events mode (by pressing Ctrl+K) then try to type using Arabic (101) keyboard layout, I get messages like these on command line and no Arabic letters are passed to my Android device:

> [server] WARN: Could not inject char u+0643
> [server] WARN: Could not inject char u+064a
> [server] WARN: Could not inject char u+0641

**Comment by AndroidDeveloperLB** [comment]
Wait, does it mean scrcpy now supports unicode?

**Comment by gaussandhisgun** [comment]
I've probably just found a way to work around this. This can be added in newer builds with some --command-line-option to enable or disable. I am not a coding genius so let me explain in my bad English:

We create an invisible always-focused text field that tracks input. Its content always defaults to, say, `>`. If any symbol was inserted or some text was pasted, device clipboard copies somewhere where it can be accessed easily, then everything after `>` is put to device clipboard, then server KEYCODE_PASTEs clipboard into the phone, and then clipboard is set back to what we saved back there. If `>` suddenly disappears (probably because it was wiped out by user who pressed Backspace), server sends KEYCODE_BACKSPACE. And all the function keys/arrows/etc are passed as it is right now if possible.

**Comment by rom1v** [comment]
@gaussandhisgun Even without "invisible text field", I did this experiment earlier: https://github.com/Genymobile/scrcpy/pull/1426

**Comment by cylee0909** [comment]
wether there is a way , we could hack 'hard keyboard' mode for input unicode chars? @rom1v

**Comment by ibrahimhajjaj** [comment]
Is there any progress on this?

**Comment by epilys** [comment]
Hello all, scrcpy was useless to me without unicode (greek) input so I made [a small GUI app](https://gist.github.com/epilys/89148373fd2fb0d1d50c941eda3ec6db) with python3 + tkinter using only standard library modules plus `xclip` and `xdotool` to **copy** the input text to system clipboard and paste it inside scrcpy with <kbd>Alt</kbd>-<kbd>v</kbd>

- `xclip` is used for setting the clipboard
- `xdotool` to search for the `scrcpy` window (set by `window_name` argument) and perform the paste shortcut

You can easily modify it to modify how the clipboard is set and how the key input is simulated (I included an example with `xte` in the comments)

Set up a macro to bring it up whenever you want unicode input using your favourite macro app.

## Demo

![demo](https://user-images.githubusercontent.com/8278356/122767977-d7ff3780-d2ab-11eb-89b0-b08bb8031fd6.png)


## Usage

```
usage: scrcpy-input.py [-h] [-m] [-nh] [-ns] window_name

This tool copies your input to system clipboard and then synchronises it with
android's clipboard and pastes it with Alt-v command. Obviously this will
overwrite your clipboard contents. Make sure you focus on the text input on
the scrcpy window before you attempt to paste.

positional arguments:

  window_name      window name to target

optional arguments:
  -h, --help       show this help message and exit
  -m, --manual     don't auto send on ctrl+Enter
  -nh, --no-hide   don't auto hide on send
  -ns, --no-strip  don't auto strip whitespace on send
```


## Source code

https://gist.github.com/epilys/89148373fd2fb0d1d50c941eda3ec6db

**Comment by AndroidDeveloperLB** [comment]
Doesn't it "contaminate" the clipboard history of the device?

Isn't it better to have some accessibility app installed, and send adb commands to it, to put the character?

**Comment by epilys** [comment]
@AndroidDeveloperLB yes it does since it syncs the clipboards. There are probably better solutions, but I made this in 2 hours without any knowledge about android or scrcpy.

**Comment by AndroidDeveloperLB** [comment]
@epilys I see.

@rom1v If you want, I can create an accessibility app. I just don't know how to make it secure, so that only via add (and maybe of a specific PC) it will allow to post the keys.
I don't want the app to be insecure, letting other apps be able to post things... 
I could perhaps offer some kind of password-like solution, so that if whoever triggers the posting of keys don't use the correct password, the app will stop working, telling the user that something weird happened...

**Comment by AlynxZhou** [comment]
> 
> 
> @rom1v What's special on Linux&USB , that sadly it works only in this scenario? Also, have you tried the solution of an accessibility app?

libusb does not work on Windows for Android HID, at least for myself.

**Comment by AndroidDeveloperLB** [comment]
@rom1v @AlynxZhou Is it possible to have this workaround instead:
Have a companion app that uses accessibility, so that when you send a command to it (with exact unicode values), it will type the real charachter 
?

**Comment by AlynxZhou** [comment]
I am not familiar with Android development...

**Comment by AndroidDeveloperLB** [comment]
@AlynxZhou What if I try it? Would you consider using it via ScrCpy (if it works) ? 
I believe it should use it only for characters that ScrCpy fail to use.

**Comment by Jacob-Darka** [comment]
Any news on making that new option `--hid-keyboard` available on windows?

**Comment by Shahin-rmz** [comment]
Hello, thanks for the awesome piece of software.
I appreciate that.
A question, today I have downloaded the debian version via apt package manager, but I can not see `--hid-keyboard` option.
Is there any cache?
Thanks

**Comment by rom1v** [comment]
Debian stable does not package the latest version. Uninstall the package and install scrcpy manually: https://github.com/Genymobile/scrcpy/blob/master/BUILD.md#simple

**Comment by Shahin-rmz** [comment]
>  install scrcpy manually: https://github.com/Genymobile/scrcpy/blob/master/BUILD.md#simple

As you said, I've done that.
![image](https://user-images.githubusercontent.com/32805117/166210982-beb51050-4882-42bd-bea8-be7c1063ea44.png)

But still can see the keyboard and can't usePersian keyboard.
The command is following right?
`scrcpy --hid-keyboard`

**Comment by rom1v** [comment]
What is the whole console output when you execute `scrcpy --hid-keyboard`?

**Comment by Shahin-rmz** [comment]
> What is the whole console output when you execute `scrcpy --hid-keyboard`?

![image](https://user-images.githubusercontent.com/32805117/166213385-d1395611-e303-4d32-917d-089b20abb06d.png)


Here you go.

**Comment by Shahin-rmz** [comment]
ps. Desktop Lubuntu 20.04
Android 11 (As you can see)

**Comment by rom1v** [comment]
OK, it looks correct, when you press a key it injects the key via HID.

Did you configure the hardware keyboard to the correct language (see end of § https://github.com/Genymobile/scrcpy#physical-keyboard-simulation-hid):

```
adb shell am start -a android.settings.HARD_KEYBOARD_SETTINGS
```
(while HID is enabled)

**Comment by mahengrui1** [comment]
recommend Gboard:
 **1** install Gboard **2** in it set non-English language with 26 key keyboard input method **3** input via scrcpy, the Gboard at Android will react as you are typing its 26 key keyboard **4** click the result via scrapy

**Comment by AndBroBeework** [comment]
Swedish letter å (u+00e5) still dont work.

```
[server] WARN: Could not inject char u+00e5
```

```
$ scrcpy -v
scrcpy 1.25 <https://github.com/Genymobile/scrcpy>

Dependencies (compiled / linked):
 - SDL: 2.26.1 / 2.26.3
 - libavcodec: 59.37.100 / 59.37.100
 - libavformat: 59.27.100 / 59.27.100
 - libavutil: 57.28.100 / 57.28.100
 - libavdevice: 59.7.100 / 59.7.100
 - libusb: - / 1.0.26

```

**Comment by pedramardakani** [comment]
> > install scrcpy manually: https://github.com/Genymobile/scrcpy/blob/master/BUILD.md#simple
> 
> As you said, I've done that. ![image](https://user-images.githubusercontent.com/32805117/166210982-beb51050-4882-42bd-bea8-be7c1063ea44.png)
> 
> But still can see the keyboard and can't usePersian keyboard. The command is following right? `scrcpy --hid-keyboard`

When I execute the program with the `--hid-keyboard` option, a notification icon comes up indicating a *Physical Keyboard* is attached. Touching it allows me to configure keyboard layouts for each keyboard available on the phone. 

For example, after clicking on `OpenBoard` with Persion layout, I chose `Farsi`, Then, simply open up a note and change the *PHONE* keyboard layout. It should write in Farsi :smiley:

**Comment by david-shu** [comment]
android手机上安装qq输入法，可以在windows下正常输入中文。

Installing QQ input method on Android phones allows for normal Chinese input under Windows.

**Comment by sambul13** [comment]
It does work with AnySoftKeyboard, once you install your chosen language pack for it on an Android phone. However, some of your language characters assigned to special characters of the physical keyboard may not print at all on PC in the phone emulator.

**Comment by rom1v** [comment]
Use `scrcpy --keyboard=uhid` since scrcpy v2.4 :rocket:

**Comment by AndroidDeveloperLB** [comment]
@rom1v Wait, this works on Windows OS too?
I will be able to write in Hebrew, for example?

**Comment by rom1v** [comment]
You tell me.

Reae the release note: https://github.com/Genymobile/scrcpy/releases/tag/v2.4

**Comment by AndroidDeveloperLB** [comment]
@rom1v I tried to use it as this command :

```
C:\android\scrcpy\scrcpy-noconsole.vbs -S --always-on-top --stay-awake --no-audio -K
```
I switched to Hebrew on Windows, and started typing in some app on the Android device via ScrCpy, but it doesn't write it. It writes in English instead.

1. Why the need of an extra parameter? In case it causes new bugs? 
2. Should I have done this differently?
3. I tried with and without it. Also tried with "--keyboard=uhid" instead. I tried on 2 different PCs. Need some environment details?

**Comment by rom1v** [comment]
Did you configure the keyboard layout on the device as explained in the doc and in the release notes?

**Comment by AndroidDeveloperLB** [comment]
@rom1v It goes there to a weird screen that I don't understand what I should do in it:

![image](https://github.com/Genymobile/scrcpy/assets/5357526/5e734b9d-7af6-4375-a8a0-f94134252add)

I tried to reach "scrcpy" there and choose Hebrew, didn't help. I chose back English-US. Still didn't help.
How can I use it?

**Comment by rom1v** [comment]
I don't know how hebrew keyboards work, but basically, this feature should work as if you plugged/connected a physical keyboard to your device.

**Comment by AndroidDeveloperLB** [comment]
@rom1v  Please explain what I should do. You can talk about any language you wish. I'm sure it should be the same.
What should I do on this screen?

**Comment by rom1v** [comment]
>  Please explain what I should do. You can talk about any language you wish.

> I tried to reach "scrcpy" there and choose Hebrew

This is the expected thing to do.

https://github.com/Genymobile/scrcpy/pull/4473#issuecomment-1888341162

**Comment by AndroidDeveloperLB** [comment]
@rom1v It seems it's not related to this screenshot, and not to the comment you've linked to.
What's missing and isn't written (please add this into the instructions), is to do this:

1. In ScrCpy, press left-ALT+K 
2. If the layouts you want are missing, reach the available layouts, and choose all languages you wish to handle. On one device they were missing so I had to choose them. On another they were already selected.
3. In order to switch languages, don't use the one of Windows OS. Instead use CTRL+Space.

Can you please somehow avoid the CTRL+Space, and just use what the (Windows) OS has? Meaning to be in sync with the current language of the OS ? It's confusing...

**Comment by AlynxZhou** [comment]
> just use what the (Windows) OS has? Meaning to be in sync with the current language of the OS

They are different input methods of independent systems, so you should never expect they are in sync, and when you press Ctrl+Space, it is likely that Windows catch it before scrcpy get it.

I think scrcpy should behave like other SDL2 games which totally disable/ignore Windows (and other OS) input methods, because when we use HID keyboard, we are using Android's input method, so the Windows input method is useless.

**Comment by AndroidDeveloperLB** [comment]
@AlynxZhou I know they are different, but Locale is a standard thing, and if they match, why not try to be in sync?

Or at least have the combination the same as on Windows? Why is it CTRL+space?
Isn't it possible at least to capture the combination that's set on the OS and use it instead?

**Comment by AlynxZhou** [comment]
> @AlynxZhou I know they are different, but Locale is a standard thing, and if they match, why not try to be in sync?
> 

Why you are assuming people have the same languages for PC and Phone? They may in different sequence, or just totally different, it needs a lot of codes to get and compare current input method.

> Or at least have the combination the same as on Windows? Why is it CTRL+space? Isn't it possible at least to capture the combination that's set on the OS and use it instead?

If you mean Super (the "Windows" key) + Space, it might be caught by Windows, so scrcpy can never get this keybinding. I previously says Ctrl+Space, but it is a mistake, I am saying Super+Space, I think this is what you want. And again, that may be caught by system before application, you may never get this keybinding in an app.

And if you still want this, please provide a API to get current keybindings for switch input method. It may still never works (because of the reason I say above), and different OS have different APIs, it is still not worthy to implement such a thing. Ignore system input method is still a better choice, because when you are using scrcpy, you really should never care about what input method is on PC.

**Comment by AlynxZhou** [comment]
> @AlynxZhou Sure it can be different settings, but it's a good thing to have, no?
> 

No, I'd like scrcpy never touches my system input method, it is not necessary for it to do so.

> About the key combination, no, I didn't mean "super"/"windows" key. I mean "CTRL" key. It's at the bottom-left corner of the keyboard, or on the left of the arrows keys. Here, found an image on the Internet to show you what CTRL key is: https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcR6csyegg19d0L9dVmZqmUiQbTyn9VkWa-YmPOkjA2t-AWKBLr9tS62Kh8FVkg32N5J9Do&usqp=CAU It's not what's set by the OS. And I'm pretty sure it is possible to capture the change of the language, because if you use a remote-control app (AnyDesk, TeamViewer, NoMachine...), they capture it fine as it passes to the controlled PC. Can't ScrCpy pass the request to change the locale of its keyboard?
> 

I know what is a "CTRL" key, if your question is why you cannot use Windows keybinding to switch input method, because which keybinding could be used to switch input method is controlled by Android here, Android does not support the keybinding you used in Windows.

The remote desktop application you mentioned can work because the target OS is still Windows so you can use Windows keybinding, but here we are targeting Android, not Windows.

scrcpy could fake key sequence, for example when get `Super+Space`, actually send `Ctrl+Space` via HID, which is what you want, but it makes what actually you get is not what actually you send, in other words, you are hiding things to user, and may leads into other conflicts.

> There is also no indication of which language is currently being set, except when you let the virtual keyboard be shown too.

Gboard will show indicator in 1. Notification Icons (with physical keyboard) 2. Space Button (when showing virtual keyboard) 3. Floating Toolbar (if you are using a recent version of it). Ask your Android input method devs to add such an indicator is a better approach, instead of using Windows' indicator for it.

**Comment by AndroidDeveloperLB** [comment]
@AlynxZhou What do you mean by "touches my system input method" ? The window is focused, and you use the key combination to change the keyboard to type in a different language.
There is even the OS setting to remember the language per app.

I talked about remote control because you said you can't capture the keys.

I didn't mention "super" key. I wrote that ScrCpy uses CTRL+Space, and the one of Window is ALT+SHIFT. I think you can change it though, and for some countries Windows decides to set it differently by default.

What things are you hiding from the user? It changes just as CTRL+space would change.
Same behavior, just different trigger.

As for indicator, that's what I wrote. It won't show an indication unless you show the virtual keyboard. Going to other places to know the current one isn't an indicator that's shown right when you are about to type. 
You wouldn't say "You can check if the smartphone has Wifi connection if you go to the Wifi screen and see it's connected", right? Such a thing appears as a real indicator in the notification/status bar. That's an indicator.

**Comment by AlynxZhou** [comment]
> the one of Window is ALT+SHIFT

Are you sure Android could switch input method with Alt+Shift?

**Comment by yume-chan** [comment]
I don't know Hebrew, but if I select Hebrew in Keyboard Layout, now every key press results in these characters

| ![image](https://github.com/Genymobile/scrcpy/assets/1330321/5d6721a1-414d-457c-827e-d8c8ce4a3747) | ![image](https://github.com/Genymobile/scrcpy/assets/1330321/bdbb6ca4-602a-4791-b9e5-81328d8f4ccd) |
|---|---|

I don't know why it's so difficult for you

**Comment by AndroidDeveloperLB** [comment]
@AlynxZhou I'm not the developer of ScrCpy and not the developer of Android framework. I'm a user of ScrCpy and I also develop apps for Android, but not of this sort.
I talked about how it works on Windows OS : ALT+SHIFT is used to switch languages, and you can choose in the settings of Windows OS that you wish each app would have it saved, so that you won't need to switch it all the time when switching between windows.

**Comment by AlynxZhou** [comment]
> I talked about how it works on Windows OS : ALT+SHIFT is used to switch languages, and you can choose in the settings of Windows OS that you wish each app would have it saved, so that you won't need to switch it all the time when switching between windows.

What if I have Japanese input method on Windows, but I don't have Japanese input method on Android?

**Comment by AndroidDeveloperLB** [comment]
@AlynxZhou Are you asking about the requesting of trying to be in sync? 
Or about the trigger of the key combination? 
If it's about the key combination, it's just triggering the switching of the language, so it goes to Android OS and there it will switch the same way it does now with ScrCpy with CTRL+Space. 

If it's about the request of trying to be in sync, I don't know what's possible so I'm not the person to ask about what are the options of what would occur.
For example, if it's all Unicode characters anyway that you can pass forward, it doesn't matter what's set on the Android OS, because you will just forward whatever is used on the Windows OS. This way, you can also send special characters, such as "😊".
My guess that as the user has a keyboard app, the options are quite restricted and it might not even work on some of them, including probably what ScrCpy has now. 

I wonder if Google will have better support for this topic now that Android Studio has its own mirroring solution. Maybe will benefit ScrCpy to support it too.
Any news for this on Android 15 , perhaps? Was it even requested, one way or another?

**Comment by AlynxZhou** [comment]
Is there anyone who really understand what on the earth his request is? I'm getting more and more confusing while reading his reply...

My opinion is scrcpy should not care about Windows input method, because we are just simulating a HID keyboard, and the Windows input method really has nothing with us because we are using Android input method. So just ignore OS input method like most SDL2 games is fine.

Anyone who understand his request please help him, I decide to only leave my opinion here.

**Comment by Suotbi** [comment]
i find the solve for this , if you want try it be me guest

**Comment by rom1v** [comment]
@KaherdinTristan use UHID keyboard: https://github.com/Genymobile/scrcpy/blob/master/doc/keyboard.md#physical-keyboard-simulation

**Comment by thejuicyroma** [comment]
Hi, I do see that the information provided by @rom1v is working for me, using a Galaxy S10. I'm just adding details explicitly for anyone who wants a step by step. Important detail is that this works on my Samsung phone using the prebuilt Samsung Keyboard. It doesn't work for me with Gboard. If you have a phone from a different maker, I can't tell you which other keyboard might work :(

1. I run scrcpy with the "--keyboard=uhid" command line argument:

C:\scrcpy-win64-v3.3.1>scrcpy --keyboard=uhid

This makes Android think I'm connecting a physical keyboard to the phone.

2. This step only needs to be performed once, not every single time: On my phone I go to Settings > General management > Samsung Keyboard settings > Languages and types > Make sure I have both "English (US) (Qwerty)" and "עברית (Qwerty)" included in the list. If not, use the 'Manage input languages' option to add both languages.

3. This step only needs to be performed once, not every single time: Settings > General management > Physical keyboard > Enable 'Show on-screen keyboard'.
 
4. This needs to happen every time I start using scrcpy: Settings > General management > Keyboard list and default > Default keyboard > Samsung Keyboard.

From this point on, when you type in scrcpy, long-press (either using your phone screen + finger, or using your mouse) on the space bar on the virtual keyboard, and select which language you want to type in. The Windows language is irrelevant, the only thing that matters is the language in the virtual keyboard.

<img width="450" height="822" alt="Image" src="https://github.com/user-attachments/assets/48c47298-9305-402e-a290-53ecf96e0e04" />

<img width="445" height="895" alt="Image" src="https://github.com/user-attachments/assets/89d6b8d0-dc6b-4890-9afc-556c31ab80d6" />


When you stop using scrcpy and want to go back to using Gboard or any other virtual keyboard, go back to step 4 and select your preferred keyboard instead of Samsung Keyboard.

Hope this helps.

---

Edit: I now learned that I can also enable "Control + Space" and/or "Shift + Space" options in order to enable switching languages directly from the keyboard without having to switch languages using the mouse or finger on the phone screen. You do this from Settings > General management > Physical keyboard > Change language shortcut. (The default "Left Alt + Shift" does not work.)

<img width="334" height="216" alt="Image" src="https://github.com/user-attachments/assets/a9a6f454-2de8-4030-99ce-1e031c91bcdf" />
